File: 20-actors/watatsuna/methods/test_plan.py
#!/usr/bin/env python3
"""Tests for the watatsuna 綿津綱 → watatsumi resilience plan (methods/plan.py).

    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest test_plan.py
    python3 test_plan.py

The load-bearing invariant (G2 + watatsumi N8): the plan can ONLY add resilience —
:lay-diverse-route / :pre-stage-repair / :monitor. There is NO interdiction/cut output by
construction. This suite proves no other plan kind can appear.
"""
from __future__ import annotations

import pathlib
import sys

try:
    from analyze import analyze, classify, load_edn
    from plan import build_plan, render_edn, render_md
except ImportError:
    from watatsuna.methods.analyze import analyze, classify, load_edn  # type: ignore
    from watatsuna.methods.plan import build_plan, render_edn, render_md  # type: ignore

_SEED = pathlib.Path(__file__).resolve().parent.parent / "data" / "seed-cable-graph.kotoba.edn"
_ALLOWED = {":lay-diverse-route", ":pre-stage-repair", ":monitor"}


def _plan():
    rows = load_edn(_SEED)
    cables, stations, links, segs, faults = classify(rows)
    a = analyze(cables, stations, links, segs, faults)
    return cables, stations, a, build_plan(cables, stations, a)


def test_plan_is_non_empty():
    *_, recs = _plan()
    assert len(recs) > 0


def test_every_plan_kind_is_resilience_only():
    *_, recs = _plan()
    kinds = {r[":plan/kind"] for r in recs}
    assert kinds <= _ALLOWED, f"non-resilience plan kind present: {kinds - _ALLOWED}"


def test_no_interdiction_kind_representable():
    *_, recs = _plan()
    for r in recs:
        k = r[":plan/kind"]
        # no cut / sever / interdict / disable verb can appear (G2 by construction)
        assert not any(bad in k for bad in ("cut", "sever", "interdict", "disable", "attack"))


def test_lay_diverse_route_targets_redundancy_gaps():
    cables, stations, a, recs = _plan()
    lay = [r for r in recs if r[":plan/kind"] == ":lay-diverse-route"]
    # there is a lay-route rec exactly where the analyzer found a single-cable landing
    assert len(lay) >= len(a["redundancy_gap"]) or len(a["redundancy_gap"]) == 0


def test_rendered_edn_marks_g2_invariant():
    *_, recs = _plan()
    edn = render_edn(recs)
    assert "redundancy + repair + monitor ONLY" in edn
    assert "interdiction" not in edn.lower() or "No interdiction" in edn


def test_rendered_md_states_watatsuna_knows_watatsumi_acts():
    *_, recs = _plan()
    md = render_md(recs)
    assert "watatsuna knows" in md and "watatsumi acts" in md
    assert "No interdiction output by construction" in md


def _run():
    fns = [v for k, v in sorted(globals().items())
           if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
    print(f"watatsuna plan.py: {len(fns)}/{len(fns)} tests passed")
    return True


if __name__ == "__main__":
    sys.exit(0 if _run() else 1)

File: 20-actors/watatsuna/methods/test_plan.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_plan.py.
(ns watatsuna.methods.test-plan
  "Tests for the watatsuna 綿津綱 → watatsumi resilience plan (methods/plan.clj).

  Load-bearing invariant (G2 + watatsumi N8): the plan can ONLY add resilience —
  :lay-diverse-route / :pre-stage-repair / :monitor. NO interdiction/cut output by construction.

  Run:  bb --classpath 20-actors 20-actors/watatsuna/methods/test_plan.clj"
  (:require [watatsuna.methods.analyze :as a]
            [watatsuna.methods.plan :as p]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.set :as set]
            [clojure.test :refer [deftest is run-tests]]))

(def ^:private this-file *file*)
(defn- seed []
  (-> this-file io/file .getAbsoluteFile .getParentFile .getParentFile
      (io/file "data" "seed-cable-graph.kotoba.edn")))
(def allowed #{:lay-diverse-route :pre-stage-repair :monitor})

(defn- plan []
  (let [g (a/classify (a/load-edn (seed)))
        an (a/analyze g)]
    [g an (p/build-plan g an)]))

(deftest plan-is-non-empty
  (let [[_ _ recs] (plan)]
    (is (pos? (count recs)))))

(deftest every-plan-kind-is-resilience-only
  (let [[_ _ recs] (plan)
        kinds (set (map :plan/kind recs))]
    (is (set/subset? kinds allowed) (str "non-resilience kind: " (set/difference kinds allowed)))))

(deftest no-interdiction-kind-representable
  (let [[_ _ recs] (plan)]
    (doseq [r recs]
      (let [k (name (:plan/kind r))]
        (is (not-any? #(str/includes? k %) ["cut" "sever" "interdict" "disable" "attack"]))))))

(deftest lay-diverse-route-targets-redundancy-gaps
  (let [[_ an recs] (plan)
        lay (filter #(= (:plan/kind %) :lay-diverse-route) recs)]
    (is (or (>= (count lay) (count (:redundancy-gap an))) (zero? (count (:redundancy-gap an)))))))

(deftest rendered-edn-marks-g2-invariant
  (let [[_ _ recs] (plan)
        edn (p/render-edn recs)]
    (is (str/includes? edn "redundancy + repair + monitor ONLY"))
    (is (not (str/includes? (str/lower-case edn) "interdiction")))))

(deftest rendered-md-states-watatsuna-knows-watatsumi-acts
  (let [[_ _ recs] (plan)
        md (p/render-md recs)]
    (is (str/includes? md "watatsuna knows"))
    (is (str/includes? md "watatsumi acts"))
    (is (str/includes? md "No interdiction output by construction"))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'watatsuna.methods.test-plan)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
